ding/reward_model/curiosity_reward_model.py

```diff
@@ -0,0 +1,201 @@
+        actions.append(action)
+    return states, next_states, actions
+
+
+class ICMNetwork(nn.Module):
+    r"""
+    Intrinsic Curiosity Model (ICM Module)
+    Implementation of:
+    [1] Curiosity-driven Exploration by Self-supervised Prediction
+    Pathak, Agrawal, Efros, and Darrell - UC Berkeley - ICML 2017.
+    https://arxiv.org/pdf/1705.05363.pdf
+
+    1) Embedding observations into a latent space
+    2) Predicting the action logit given two consecutive embedded observations
+    3) Predicting the next embedded obs, given the embeded former observation and action
+    """
+
+    def __init__(self, obs_shape: Union[int, SequenceType], hidden_size_list: SequenceType, action_shape: int) -> None:
+        super(ICMNetwork, self).__init__()
+        if isinstance(obs_shape, int) or len(obs_shape) == 1:
+            self.feature = FCEncoder(obs_shape, hidden_size_list)
+        elif len(obs_shape) == 3:
+            self.feature = ConvEncoder(obs_shape, hidden_size_list)
+        else:
+            raise KeyError(
+                "not support obs_shape for pre-defined encoder: {}, please customize your own ICM model".
+                format(obs_shape)
+            )
+        self.action_shape = action_shape
+        feature_output = hidden_size_list[-1]
+        self.inverse_net = nn.Sequential(nn.Linear(feature_output * 2, 512), nn.ReLU(), nn.Linear(512, action_shape))
+        self.residual = [
+            nn.Sequential(
+                nn.Linear(action_shape + 512, 512),
+                nn.LeakyReLU(),
+                nn.Linear(512, 512),
+            ) for _ in range(8)
+        ]
+        self.forward_net_1 = nn.Sequential(nn.Linear(action_shape + feature_output, 512), nn.LeakyReLU())
+        self.forward_net_2 = nn.Sequential(nn.Linear(action_shape + 512, feature_output), )
+
+    def forward(self, state: torch.Tensor, next_state: torch.Tensor,
+                action_long: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        r"""
+        Overview:
+            Use observation, next_observation and action to genearte ICM module
+            Parameter updates with ICMNetwork forward setup.
+        Arguments:
+            - state (:obj:`torch.Tensor`):
+                The current state batch, (B,N=observation_size)``.
+            - next_state (:obj:`torch.Tensor`):
+                The next state batch, (B,N=observation_size)``.
+            - action_long (:obj:`torch.Tensor`):
+                The action, (B,M=action_size)``.
+        Returns:
+            - real_next_state_feature (:obj:`torch.Tensor`):
+                Run with the encoder. Return the real next_state's embedded feature.
+            - pred_next_state_feature (:obj:`torch.Tensor`):
+                Run with the encoder and residual network. Return the predicted next_state's embedded feature.
+            - pred_action_logit (:obj:`Dict`):
+                Run with the encoder. Return the predicted action logit.
+        """
```

Comment:
add shapes comments for this method, like [this](https://github.com/opendilab/DI-engine/blob/main/ding/model/template/q_learning.py#L76)